### PR TITLE
Us024 registration system

### PR DESCRIPTION
**Pull request by mtslech**
### 🛠 Changes being made

Here give examples of the changes you've made in this pull request. Include an itemized list if you can. It'll help the reviewer

### ✨ Definition of Done (all the points need to be filled out before merging)

- [x] Meets Acceptance Criteria
- [x] Tested
- [x] Code Review made by team 2 members
- [x] UX approved
- [x] Product Owner Approved

### 🧪 Non-Functional Requirements

- [x] Automated tests should cover acceptance criterias (Unit tests, Integration/API/Component tests)
- [x] All input must be validated (client side & API)
- [x] Sensitive data should not be stored in the URL
- [x] No secrets in codesource

### CODE REVIEW

File: product-backend/src/main/java/org/kainos/ea/db/AuthDao.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.kainos.ea.model.User;
+import org.mindrot.jbcrypt.BCrypt;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+
+public class AuthDao {
+    private DatabaseConnector databaseConnector = new DatabaseConnector();
```
Comment (Fluffyguy611) [inline product-backend/src/main/java/org/kainos/ea/db/AuthDao.java]: Should be created in construction in Service class

File: product-web-app/views/register.html [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+<body>
+    <header class="bg-success text-light">
+        <div class="container">
+            <div class="row">
+                <div class="col-md-12">
+                    <h1>Register</h1>
+                </div>
+            </div>
+        </div>
+    </header>
```
Comment (Fluffyguy611) [inline product-web-app/views/register.html]: Add parent.html extension: 
see readme.md on teams

**Review by emiliabrzezinska — APPROVED** [review]

File: product-backend/src/main/java/org/kainos/ea/controller/AuthController.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    @POST
+    @Path("/auth/register")
+    @Produces(MediaType.APPLICATION_JSON)
+    @Consumes(MediaType.APPLICATION_JSON)
+    public Response register(User user) {
+        try {
+            return Response.ok(authService.register(user)).build();
+        } catch (FailedToRegisterException | InvalidEmailException | InvalidPasswordException |
+                 InvalidRoleIdException err) {
+            System.err.println(err.getMessage());
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/controller/AuthController.java]: Use logger see here
https://github.com/ljercha/agile-project/blob/main/product-backend/src/main/java/org/kainos/ea/controller/ProductController.java

File: product-backend/src/main/java/org/kainos/ea/db/AuthDao.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+        st.setInt(3, user.getRoleId());
+        st.executeUpdate();
+
+        return user.getEmail();
+    }
+
+    private String hashPassword(String plainPassword) {
+        return BCrypt.hashpw(plainPassword, BCrypt.gensalt());
+    }
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/db/AuthDao.java]: That method should be a part of the AuthService
```suggestion
    private String hashPassword(String plainPassword) {
        return BCrypt.hashpw(plainPassword, BCrypt.gensalt());
    }
```

File: product-backend/src/main/java/org/kainos/ea/db/DatabaseConnector.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+public class DatabaseConnector {
+    private static Connection conn;
+
+    public static Connection getConnection() throws SQLException {
+        String user, password, host, name;
+        if (conn != null && !conn.isClosed()) {
+            return conn;
+        }
+
+        try (FileInputStream propsStream = new FileInputStream("db.properties")) {
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/db/DatabaseConnector.java]: Use environment variables - see in example project or take a look at Adrian or Bartlomiej PR (you can just copy file)

File: product-backend/src/main/java/org/kainos/ea/service/AuthService.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+            if (!authValidator.isRoleIdValid(user.getRoleId())) {
+                throw new InvalidRoleIdException();
+            }
+
+            authDao.registerUser(user, databaseConnector.getConnection());
+            return user.getEmail();
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
+            throw new FailedToRegisterException();
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/service/AuthService.java:47]: Use logger see in examples

File: product-backend/src/main/java/org/kainos/ea/service/AuthService.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            }
+
+            if (!authValidator.isRoleIdValid(user.getRoleId())) {
+                throw new InvalidRoleIdException();
+            }
+
+            authDao.registerUser(user, databaseConnector.getConnection());
+            return user.getEmail();
+        } catch (SQLException e) {
+            System.err.println(e.getMessage());
```
Comment (ljercha) [inline product-backend/src/main/java/org/kainos/ea/service/AuthService.java]: use logger

File: product-backend/src/test/java/AuthServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+import org.kainos.ea.service.AuthService;
+import org.kainos.ea.service.AuthValidator;
+
+import java.sql.SQLException;
+
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+class AuthServiceTest {
```
Comment (ljercha) [inline product-backend/src/test/java/AuthServiceTest.java:14]: No happy path

When_User_IsValid_Expect_EmailReturned

 you should verify
-  authDao mock was called
-  email have been returned

File: product-web-app/service/authService.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    const validateEmailError = this.authValidator.validateEmail(user.email);
+    if (validateEmailError) {
+      throw new Error(validateEmailError);
+    }
+
+    try {
+      const response = await axios.post('http://localhost:8080/api/auth/register', user);
+      console.log('User registered:', response.data);
```
Comment (ljercha) [inline product-web-app/service/authService.ts]: user logger

File: product-web-app/service/authService.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import AuthValidator from './authValidator.js';
+
+export default class AuthService {
+  private authValidator: AuthValidator;
+
+  constructor(authValidator: AuthValidator) {
+    this.authValidator = authValidator;
+  }
+
+  async register(user: User): Promise<void> {
```
Comment (ljercha) [inline product-web-app/service/authService.ts]: Add repeat password field on the frontend and validate it here

File: product-web-app/service/authService.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+      throw new Error(validatePasswordError);
+    }
+
+    const validateEmailError = this.authValidator.validateEmail(user.email);
+    if (validateEmailError) {
+      throw new Error(validateEmailError);
+    }
+
+    try {
+      const response = await axios.post('http://localhost:8080/api/auth/register', user);
```
Comment (ljercha) [inline product-web-app/service/authService.ts]: remove the base url, setup it in app.ts see in example project or in other PR's
https://github.com/ljercha/agile-project/blob/fd328a0b0e3de1bc7714810cf5d2c308a447f879/product-web-app/app.ts#L32C34-L32C34

File: product-web-app/views/register.html [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+                        {{ errormessage }}
+                    </div>
+                {% endif %}
+                <form action="/register" method="post">
+                    <div class="form-group">
+                        <label for="email">Email</label>
+                        <input type="text" id="email" name="email" class="form-control" value="{{ email }}" required>
+                    </div>
+                    <div class="form-group">
+                        <label for="password">Password</label>
```
Comment (ljercha) [inline product-web-app/views/register.html]: Add repeat password field

File: product-web-app/views/register.html [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+        <section id="form" class="bg-light py-5">
+            <div class="container">
+                {% if errormessage %}
+                    <div class="alert alert-danger" role="alert">
+                        {{ errormessage }}
+                    </div>
+                {% endif %}
+                <form action="/register" method="post">
+                    <div class="form-group">
+                        <label for="email">Email</label>
```
Comment (ljercha) [inline product-web-app/views/register.html]: Add an information about the fields requirements e.g. email must be in kainos domain, password must have special character etc.

File: product-web-app/service/authValidator.ts [hunk 14]
```diff
@@ -0,0 +1,34 @@
+export default class AuthValidator {
```
Comment (ljercha) [inline product-web-app/service/authValidator.ts:1]: missing unit tests

File: product-web-app/service/authService.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import axios from 'axios';
+import User from '../model/user.js';
+import AuthValidator from './authValidator.js';
+
+export default class AuthService {
```
Comment (ljercha) [inline product-web-app/service/authService.ts:7]: missing unit tests

File: product-web-app/app.ts [hunk 16]
```diff
@@ -44,6 +45,16 @@ app.listen(3000, () => {
   logger.info('Server listening on port 3000');
 });
 
+
+const authController = new AuthController();
+
+app.post('/', (req: Request, res: Response) => {
+  res.redirect('/auth/register');
+ });
```
Comment (ljercha) [inline product-web-app/app.ts]: what is that

File: product-web-app/package.json [hunk 17]
```diff
@@ -15,6 +15,7 @@
     "test": "ts-mocha"
   },
   "dependencies": {
+    "all": "^0.0.0",
```
Comment (ljercha) [inline product-web-app/package.json]: WTF

File: product-web-app/views/register.html [hunk 18]
```diff
@@ -0,0 +1,54 @@
+<!DOCTYPE html>
+<html>
+<head>
```
Comment (ljercha) [inline product-web-app/views/register.html]: use template from main

**Review by ljercha — APPROVED** [review]

**Pull request merged**